google-cloud-core/src/main/java/com/google/cloud/testing/BaseEmulatorHelper.java

```diff
@@ -104,16 +109,50 @@ protected final void startProcess(String blockUntilOutput)
   }
 
   /**
-   * Stops the local service's subprocess and any possible thread listening for its output.
+   * Waits for the local service's subprocess to terminate,
+   * and stop any possible thread listening for its output.
    */
-  protected final void stopProcess() throws IOException, InterruptedException {
+  protected final int waitForProcess(Duration timeout) throws IOException, InterruptedException, TimeoutException {
     if (blockingProcessReader != null) {
       blockingProcessReader.terminate();
       blockingProcessReader = null;
     }
     if (activeRunner != null) {
-      activeRunner.stop();
+      int exitCode = activeRunner.waitFor(timeout);
       activeRunner = null;
+      return exitCode;
+    }
+    return 0;
+  }
+
+  private static int waitForProcess(final Process process, Duration timeout) throws InterruptedException, TimeoutException {
+    if (process == null) {
+      return 0;
+    }
+
+    final SettableFuture<Integer> exitValue = SettableFuture.create();
+
+    Thread waiter = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          exitValue.set(process.waitFor());
+        } catch (InterruptedException e) {
+          exitValue.setException(e);
+        }
+      }
+    });
+    waiter.run();
```

Comment:
I think Codacy caught a bug here - won't this execute your "threaded" code in the thread calling `run()`?